contracts/modules/Experimental/TransferManager/LockupVolumeRestrictionTM.sol

```diff
@@ -229,164 +238,208 @@ contract LockupVolumeRestrictionTM is ITransferManager {
+            }
+            
+    }
 
     /**
      * @notice Parameter checking function for creating or editing a lockup.  This function will cause an exception if any of the parameters are bad.
-     * @param lockUpPeriodSeconds Total period of lockup (seconds)
-     * @param releaseFrequencySeconds How often to release a tranche of tokens (seconds)
-     * @param totalAmount Total amount of locked up tokens
+     * @param _userAddress Address whom lockup is being applied
+     * @param _lockupAmount Amount that needs to be locked
+     * @param _lockUpPeriodSeconds Total period of lockup (seconds)
+     * @param _releaseFrequencySeconds How often to release a tranche of tokens (seconds)
      */
-    function _checkLockUpParams(uint lockUpPeriodSeconds, uint releaseFrequencySeconds, uint totalAmount) internal view {
-        require(lockUpPeriodSeconds != 0, "lockUpPeriodSeconds cannot be zero");
-        require(releaseFrequencySeconds != 0, "releaseFrequencySeconds cannot be zero");
-        require(totalAmount != 0, "totalAmount cannot be zero");
+    function _checkLockUpParams(
+        address _userAddress,
+        uint256 _lockupAmount,
+        uint256 _lockUpPeriodSeconds,
+        uint256 _releaseFrequencySeconds
+    ) 
+        internal 
+        view 
+    {   
+        require(_userAddress != address(0), "Invalid address");
+        require(_lockUpPeriodSeconds != 0, "lockUpPeriodSeconds cannot be zero");
+        require(_releaseFrequencySeconds != 0, "releaseFrequencySeconds cannot be zero");
+        require(_lockupAmount != 0, "lockupAmount cannot be zero");
 
         // check that the total amount to be released isn't too granular
         require(
-            totalAmount % ISecurityToken(securityToken).granularity() == 0,
+            _lockupAmount % ISecurityToken(securityToken).granularity() == 0,
             "The total amount to be released is more granular than allowed by the token"
         );
 
         // check that releaseFrequencySeconds evenly divides lockUpPeriodSeconds
         require(
-            lockUpPeriodSeconds % releaseFrequencySeconds == 0,
+            _lockUpPeriodSeconds % _releaseFrequencySeconds == 0,
             "lockUpPeriodSeconds must be evenly divisible by releaseFrequencySeconds"
         );
 
-        // check that totalPeriods evenly divides totalAmount
-        uint totalPeriods = lockUpPeriodSeconds.div(releaseFrequencySeconds);
+        // make sure the amount to be released per period is not too granular for the token
+        uint256 totalPeriods = _lockUpPeriodSeconds.div(_releaseFrequencySeconds);
+        uint256 amountPerPeriod = _lockupAmount.div(totalPeriods);
         require(
-            totalAmount % totalPeriods == 0,
-            "The total amount being locked up must be evenly divisible by the number of total periods"
+            amountPerPeriod.mul(totalPeriods) == _lockupAmount,
+            "lockup amount should be completely divisible by the amount per period"
```

Comment:
Wondering how necessary this check is? Suppose that the amountPerPeriod is not a whole number (e.g. 10.5 tokens released per period) this would just effectively mean that the user has to keep a balance of 11 tokens to satisfy the clause.

Reply:
Without this check, rounding off can cause tokens to be locked forever. Consider this example:
Granularity of token = 1
Total locked amount = 10
Number of periods = 6

This means, the amount per lock period = 10/6 = 1.67 = 1.
unlocked amount even after all periods are passed = 6*1 = 6
10 - 6 = 4 tokens are locked forever.

4 is a very small number as 1 token is 10**18 but it can cause amount mismatch and confusion as it will make transactions revert when user will try to transfer all of locked tokens.

That being said, I don't think we should be checking against granularity for amount released per period. We are never taking tokens away from the guy so the granularity issue is not a big problem for individual release amounts.

Reply:
I guess this would need to be combined with the comment above that ignores any lockups which are past their end time. I feel like we should be doing this anyway for gas purposes.